packages/dockest/src/run/waitForServices/index.ts

```diff
@@ -76,13 +67,42 @@ export const waitForServices = async ({
     await joinBridgeNetwork({ containerId: hostname, alias: DOCKEST_HOST_ADDRESS })
   }
 
-  for (const runner of Object.values(runners)) {
-    if (runInBand) {
-      await waitForRunner({ runner })
+  const dependencyGraph: Array<[string, string | undefined]> = []
+
+  const walkRunner = (runner: Runner) => {
+    if (mutables.runnerLookupMap.has(runner.serviceName)) {
+      return
+    }
+    mutables.runnerLookupMap.set(runner.serviceName, runner)
+
+    if (runner.dependsOn.length === 0) {
+      dependencyGraph.push([runner.serviceName, undefined])
     } else {
-      setupPromises.push(waitForRunner({ runner }))
+      for (const dependencyRunner of runner.dependsOn) {
+        dependencyGraph.push([dependencyRunner.serviceName, runner.serviceName])
+        walkRunner(dependencyRunner)
+      }
     }
   }
 
-  await Promise.all(setupPromises)
+  for (const runner of Object.values(mutables.runners)) {
+    walkRunner(runner)
+  }
+
+  if (runInBand) {
+    const ordered: Array<string> = toposort(dependencyGraph).filter(value => value !== undefined)
+
+    const teardownOrder = ordered.slice(0).reverse()
+    mutables.teardownOrder = teardownOrder
+
+    for (const serviceName of ordered) {
+      const runner = mutables.runnerLookupMap.get(serviceName)
+      if (!runner) {
+        throw new DockestError('Unexpected error. Runner could not be found.')
+      }
+      await waitForRunner({ runner })
+    }
+  } else {
+    await Promise.all(Array.from(mutables.runnerLookupMap.values()).map(runner => waitForRunner({ runner })))
+  }
```

Comment:
Right now in `runInBand`, all the services are started one after another (as determined by `toposort`). But instead, we could generate a dependency list with the type of `Array<Array<string>>` and then start all those services of each `Array<string>` element in parallel.

E.g. the following depends on config
```
A depends on B, C
B depends on D
C depends on D
D depends on nothing
```

will result in the following dependency list:

`[["D"], ["B", "C"], ["A"]]`

```
Start D - Wait until D has passed the readiness check
Start B and C - Wait until B and C have passed the readiness check
Start A - Wait until A has passed the readiness check
```

```